### PR TITLE
Integrate MKLDNN Conv1d and support 3d layout

### PR DESCRIPTION
**Pull request by xinyu-intel**
## Description ##
This PR aims to integrate MKLDNN Conv1d and enable 3d layout for Conv and Activation.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] MKLDNN Conv1d
- [x] MKLDNN 3d layout for Conv and Activation

@pengzhao-intel @TaoLv

### CODE REVIEW

**Comment by TaoLv** [comment]
Related issues: https://github.com/apache/incubator-mxnet/issues/11906, https://github.com/apache/incubator-mxnet/issues/11161

**Comment by TaoLv** [comment]
@mxnet-label-bot add [MKLDNN, Operator, pr-awaiting-review]

File: src/operator/nn/activation.cc [hunk 1]
```diff
@@ -97,9 +97,10 @@ static void ActivationComputeExCPU(const nnvm::NodeAttrs& attrs,
                                    const std::vector<NDArray>& inputs,
                                    const std::vector<OpReqType>& req,
                                    const std::vector<NDArray>& outputs) {
+  const ActivationParam& param = nnvm::get<ActivationParam>(attrs.parsed);
   CHECK_EQ(inputs.size(), 1U);
   CHECK_EQ(outputs.size(), 1U);
-  if (SupportMKLDNN(inputs[0])) {
+  if (SupportMKLDNNAct(param, inputs[0])) {
```
Comment (TaoLv) [inline src/operator/nn/activation.cc]: Please confirm if changes also need be applied to L139 and L150.

File: src/operator/nn/mkldnn/mkldnn_act.cc [hunk 2]
```diff
@@ -49,6 +49,17 @@ bool SupportMKLDNNAct(const ActivationParam& param) {
       || param.act_type == activation::kTanh;
 }
 
+bool SupportMKLDNNAct(const ActivationParam& param, const NDArray &input) {
+  if ((input.shape().ndim() < 1) ||
+      (input.shape().ndim() > 4) ||
+      (input.dtype() != mshadow::kFloat32))
+    return false;
+  return param.act_type == activation::kReLU
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_act.cc]: return SupportMKLDNNAct(param);

File: src/operator/nn/mkldnn/mkldnn_act.cc [hunk 3]
```diff
@@ -49,6 +49,17 @@ bool SupportMKLDNNAct(const ActivationParam& param) {
       || param.act_type == activation::kTanh;
 }
 
+bool SupportMKLDNNAct(const ActivationParam& param, const NDArray &input) {
+  if ((input.shape().ndim() < 1) ||
+      (input.shape().ndim() > 4) ||
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_act.cc]: Please add comments to this dimension check.

File: src/operator/nn/mkldnn/mkldnn_base-inl.h [hunk 4]
```diff
@@ -253,14 +254,23 @@ inline static mkldnn::memory::desc GetWeightDesc(const NDArray &arr,
...
   } else {
-    CHECK_EQ(arr.shape().ndim(), 4U);
-    mkldnn::memory::dims tz = mkldnn::memory::dims{ num_groups,
-      static_cast<int>(arr.shape()[0] / num_groups),
-      static_cast<int>(arr.shape()[1]),
-      static_cast<int>(arr.shape()[2]),
-      static_cast<int>(arr.shape()[3])};
-    return mkldnn::memory::desc{tz, get_mkldnn_type(arr.dtype()),
-                                mkldnn::memory::format::any};
+    CHECK((arr.shape().ndim() == 3) || (arr.shape().ndim() == 4));
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_base-inl.h]: Add error message if the CHECK failed.

File: src/operator/nn/mkldnn/mkldnn_base-inl.h [hunk 5]
```diff
@@ -253,14 +254,23 @@ inline static mkldnn::memory::desc GetWeightDesc(const NDArray &arr,
...
+    if (arr.shape().ndim() == 3) {
+      mkldnn::memory::dims tz = mkldnn::memory::dims{ num_groups,
+        static_cast<int>(arr.shape()[0] / num_groups),
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_base-inl.h]: fix indents.

File: src/operator/nn/mkldnn/mkldnn_base-inl.h [hunk 6]
```diff
@@ -253,14 +254,23 @@ inline static mkldnn::memory::desc GetWeightDesc(const NDArray &arr,
...
+        static_cast<int>(arr.shape()[1]),
+        static_cast<int>(arr.shape()[2])};
+      return mkldnn::memory::desc{tz, get_mkldnn_type(arr.dtype()),
+                                  mkldnn::memory::format::any};
+    } else {
+      mkldnn::memory::dims tz = mkldnn::memory::dims{ num_groups,
+        static_cast<int>(arr.shape()[0] / num_groups),
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_base-inl.h]: fix indents.

File: src/operator/nn/mkldnn/mkldnn_base-inl.h (hunk 5 above)
Comment (TaoLv) [reply src/operator/nn/mkldnn/mkldnn_base-inl.h]: use N, C, H, W instead of 0, 1, 2, 3.

File: src/operator/nn/mkldnn/mkldnn_base.cc [hunk 7]
```diff
@@ -238,39 +238,44 @@ const mkldnn::memory *GetWeights(const NDArray &arr,
...
-          static_cast<int>(arr.shape()[2]), static_cast<int>(arr.shape()[3])};
-    mkldnn::memory::desc md =
-        mkldnn::memory::desc{tz, type, mkldnn::memory::format::oihw};
-    mkldnn::memory::primitive_desc pd =
-        mkldnn::memory::primitive_desc{md, engine};
-    mem = arr.GetMKLDNNData(pd);
+    format = mkldnn::memory::format::oi;
+  } else if (arr.shape().ndim() == 3) {
+    tz = num_groups > 1 ? mkldnn::memory::dims{ num_groups,
+      static_cast<int>(arr.shape()[0] / num_groups),
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_base.cc]: fix indents.

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 8]
```diff
@@ -51,15 +54,23 @@ mkldnn::convolution_forward::primitive_desc GetConvFwdImpl(
...
+    padding[0] = param.conv_param.pad[0];
+  } else if (param.conv_param.kernel.ndim() == 2) {
+    CHECK_GE(param.conv_param.stride.ndim(), 2U);
+    CHECK_GE(param.conv_param.pad.ndim(), 2U);
+    CHECK_GE(param.conv_param.dilate.ndim(), 2U);
+    strides[0] = param.conv_param.stride[0];
+    strides[1] = param.conv_param.stride[1];
+    padding[0] = param.conv_param.pad[0];
+    padding[1] = param.conv_param.pad[1];
+  }
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc]: else?

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 9]
```diff
@@ -113,9 +124,13 @@ mkldnn::convolution_forward::primitive_desc GetConvFwdImpl(
...
-    mkldnn::memory::dims dilates{0, 0};
-    dilates[0] = param.conv_param.dilate[0] - 1;
-    dilates[1] = param.conv_param.dilate[1] - 1;
+    mkldnn::memory::dims dilates(param.conv_param.kernel.ndim());
+    if (param.conv_param.dilate.ndim() == 1) {
+      dilates[0] = param.conv_param.dilate[0] - 1;
+    } else if (param.conv_param.dilate.ndim() == 2) {
+      dilates[0] = param.conv_param.dilate[0] - 1;
+      dilates[1] = param.conv_param.dilate[1] - 1;
+    }
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc]: else?

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 10]
```diff
@@ -151,15 +166,23 @@ static mkldnn::convolution_backward_data::primitive_desc GetConvBwdData(
...
+    padding[0] = param.pad[0];
+  } else if (param.kernel.ndim() == 2) {
+    CHECK_GE(param.stride.ndim(), 2U);
+    CHECK_GE(param.pad.ndim(), 2U);
+    CHECK_GE(param.dilate.ndim(), 2U);
+    strides[0] = param.stride[0];
+    strides[1] = param.stride[1];
+    padding[0] = param.pad[0];
+    padding[1] = param.pad[1];
+  }
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc]: else?

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 11]
```diff
@@ -239,9 +274,13 @@ static mkldnn::convolution_backward_weights::primitive_desc GetConvBwdWeights(
...
-    mkldnn::memory::dims dilates{0, 0};
-    dilates[0] = param.dilate[0] - 1;
-    dilates[1] = param.dilate[1] - 1;
+    mkldnn::memory::dims dilates(param.kernel.ndim());
+    if (param.dilate.ndim() == 1) {
+      dilates[0] = param.dilate[0] - 1;
+    } else if (param.dilate.ndim() == 2) {
+      dilates[0] = param.dilate[0] - 1;
+      dilates[1] = param.dilate[1] - 1;
+    }
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc]: else

**Comment by xinyu-intel** [comment]
@TaoLv address Tao's comments. These changes are already covered by `test_operator:test_convolution_grouping` and `test_operator:test_activation`. I check if they run into MKL-DNN kernel by checking the mkldnn_verbose.

**Comment by TaoLv** [comment]
Do we have unit test for 1D Convolution without grouping? Please also make sure that this change works well for quantized Convolution (which I think doesn't support 1D yet).

**Comment by xinyu-intel** [comment]
Address Tao's comments. I've skip conv1d convolution on both C level and Python level. @ZhennanQin Please help take a look. Maybe skip all non-4d data layout in quantization.py is not a good choice.

File: python/mxnet/contrib/quantization.py [hunk 12]
```diff
@@ -488,6 +488,9 @@ def quantize_model(sym, arg_params, aux_params,
         A tuple of quantized symbol, quantized arg_params, and aux_params.
     -------
     """
+    if ctx == cpu(0) and len(calib_data.provide_data[0].shape) != 3:
+        raise ValueError('MKL-DNN quantized OPs temporary support 4d layout.')
```
Comment (ZhennanQin) [inline python/mxnet/contrib/quantization.py]: Please don't check calib_data as quantization flow support non-calib mode.

Comment (xinyu-intel) [reply python/mxnet/contrib/quantization.py]: done

**Comment by ZhennanQin** [comment]
LGTM.

**Comment by TaoLv** [comment]
Seems there was problem with CI. Do you mind re-triggering it with an empty commit? @xinyu-intel 
Ping @zheng-da for review. Thank you.

File: src/operator/nn/mkldnn/mkldnn_base-inl.h [hunk 13]
```diff
@@ -253,14 +254,23 @@ inline static mkldnn::memory::desc GetWeightDesc(const NDArray &arr,
...
+          num_groups, static_cast<int>(arr.shape()[N] / num_groups),
+          static_cast<int>(arr.shape()[C]), static_cast<int>(arr.shape()[H])};
+      return mkldnn::memory::desc{tz, get_mkldnn_type(arr.dtype()),
+                                  mkldnn::memory::format::any};
+    } else {
+      mkldnn::memory::dims tz = mkldnn::memory::dims{
+          num_groups, static_cast<int>(arr.shape()[N] / num_groups),
+          static_cast<int>(arr.shape()[C]), static_cast<int>(arr.shape()[H]),
+          static_cast<int>(arr.shape()[W])};
+      return mkldnn::memory::desc{tz, get_mkldnn_type(arr.dtype()),
```
Comment (pengzhao-intel) [inline src/operator/nn/mkldnn/mkldnn_base-inl.h]: This line is the common part of dim3 and dim4, right?

File: src/operator/nn/mkldnn/mkldnn_base-inl.h [hunk 14]
```diff
@@ -253,14 +254,23 @@ inline static mkldnn::memory::desc GetWeightDesc(const NDArray &arr,
...
   } else {
-    CHECK_EQ(arr.shape().ndim(), 4U);
-    mkldnn::memory::dims tz = mkldnn::memory::dims{ num_groups,
-      static_cast<int>(arr.shape()[0] / num_groups),
-      static_cast<int>(arr.shape()[1]),
-      static_cast<int>(arr.shape()[2]),
-      static_cast<int>(arr.shape()[3])};
-    return mkldnn::memory::desc{tz, get_mkldnn_type(arr.dtype()),
-                                mkldnn::memory::format::any};
+    CHECK((arr.shape().ndim() == 3) || (arr.shape().ndim() == 4))
```
Comment (pengzhao-intel) [inline src/operator/nn/mkldnn/mkldnn_base-inl.h]: Use a variable to save the value of `arr.shape().ndim()` to avoid mutiple time call

**Comment by xinyu-intel** [comment]
@pengzhao-intel Conv1d will fall back to native cpu implement.

100 iterations total time(ms) on Xeon Skylake 8180 1 socket:

|shape|before opt|after opt|speedup|
|:---:|:---:|:---:|:---:|
|(1,256,200)|715.47|73.88|9.68x|
|(1,1024,512)|1970.15|101.06|19.49x|
|(64,1024,512)|131312.48|4196.24|31.29x|

I've add 1d,3d,4d data shape to activation test. Regarding quantized conv, it will now return error when the data shape is 3d and users should exclude this layer:

```
  CHECK_EQ(param.full_conv_param.conv_param.kernel.ndim(), 2U)
      << "MKL-DNN only supports quantized conv2d.";
```

**Comment by xinyu-intel** [comment]
@zheng-da Please take a look. Thanks!

**Comment by TaoLv** [comment]
@xinyu-intel Thank you for the contribution. Now merging.

**Comment by bputrycz** [comment]
I noticed a very nice improvement with this change.
So, thank you.

Still, for my use-case: conv1d, small batch size, small channel dimension, long sequence length,
I don't see much improvement with more cores added to the computations.

The simplified snippet to reproduce (conv1d.py):
```
import mxnet as mx
from mxnet import gluon, nd

from mxnet import profiler
profiler.set_config(profile_all=True, aggregate_stats=True, filename='profile_output.json')
 
channels = 64

net = gluon.nn.Sequential()

conv = gluon.nn.Conv1D(channels, 4, padding=1)
act = gluon.nn.Activation('sigmoid')

for i in range(3):
    net.add(conv)
    net.add(act)

net.initialize()

data = nd.random.uniform(shape=(1, channels, 2**16))

# Warm-up
y = net(data)
nd.waitall()

profiler.set_state('run')
for i in range(10):
    y = net(data)
    nd.waitall()
profiler.set_state('stop')

print(profiler.dumps())
```
When run on a host with a lot of cores (AWS c4.8xlarge) results in:
```
$ OMP_NUM_THREADS=1 python conv1d.py | grep "Convolution\|Activation"
Activation                             60         244.8700           3.3960           9.7510           4.0812
Convolution                            60        1648.7010          24.9280          40.6520          27.4783
$ OMP_NUM_THREADS=2 python conv1d.py | grep "Convolution\|Activation"
Activation                             60         127.4460           1.6600           5.4070           2.1241
Convolution                            60         866.8680          12.6670          22.9810          14.4478
$ OMP_NUM_THREADS=4 python conv1d.py | grep "Convolution\|Activation"
Activation                             60          65.3190           0.8940           2.9280           1.0886
Convolution                            60         854.2900          12.6230          20.3230          14.2382
```

There is no improvement when number of threads is increased to 4, or more.

Playing more with this example, for higher 'channels' value, it starts to be a little better parallelizable.
So, it seems the parallelization is done only per a single sequence "point". Is it the case?
But, it seems quite natural to parallelize also along the sequence, especially when it is long - different threads doing a different part of the sequence.
Then, parallelization should scale linearly with number of threads.
Isn't it done like that?

Bartosz

**Comment by pengzhao-intel** [comment]
Thanks @bputrycz really thanks for trying the new MKLDNN API and give us the very useful feedback.
Nice analysis and reproducible examples. I will contact Intel MKL-DNN team to see how to fix it.

In a short term, you can try to launch multiple instances (processor) and each one binds to 2/4 cores to get the max throughput in case this is the inference case.

**Comment by pengzhao-intel** [comment]
@TaoLv is following this issue and will back to you with more details soon.

**Comment by TaoLv** [comment]
@bputrycz Cannot reproduce the problem. The code snippet you provided scales well on my machine:
```
$ OMP_NUM_THREADS=1 python conv1d.py | grep "Convolution\|Activation"
Activation                             60         183.5660           2.6600           6.1490           3.0594
Convolution                            60        1047.5291          14.2060          26.2510          17.4588
$
$ OMP_NUM_THREADS=2 python conv1d.py | grep "Convolution\|Activation"
Activation                             60         103.8990           1.4330           3.5280           1.7316
Convolution                            60         550.4830           7.5440          13.9890           9.1747
$
$ OMP_NUM_THREADS=4 python conv1d.py | grep "Convolution\|Activation"
Activation                             60          57.5180           0.7640           2.4380           0.9586
Convolution                            60         290.6110           3.8830           7.8330           4.8435
$
$ OMP_NUM_THREADS=8 python conv1d.py | grep "Convolution\|Activation"
Activation                             60          39.8030           0.4470           1.8200           0.6634
Convolution                            60         179.9880           2.2860           5.4140           2.9998
```

Have you ever tried the cpu affinity before run multi-thread case?
```
export KMP_AFFINITY=granularity=fine,compact
```

**Comment by TaoLv** [comment]
Ummm, just notice that you're using c4.8xlarge which I think have no AVX512. Will have another try and come back to you later.

**Comment by TaoLv** [comment]
Confirmed this problem exists on machine without AVX512 as threading optimization for 1D Convolution is made for AVX512 only. Would you mind having a try on AWS c5? @bputrycz

**Comment by bputrycz** [comment]
Yes. I confirm.
It scales much better on AWS c5.

Thank you very much.

**Comment by pengzhao-intel** [comment]
> Yes. I confirm.
> It scales much better on AWS c5.
> 
> Thank you very much.

It's good to see the problem can be resolved on AWS c5. 
Feel free to ping us if you have any questions or issues :)
We will document this behavior in [MKLDNN README](https://github.com/apache/incubator-mxnet/blob/master/MKLDNN_README.md) @xinyu-intel

**Comment by TaoLv** [comment]
@bputrycz Also feel free to let me know if the performance on c4 is critical for you.

**Comment by bputrycz** [comment]
@TaoLv Performance on c4 is not critical for me, as for now.

**Pull request merged**